### PR TITLE
applications: nrf5340_audio: Move data_fifo

### PR DESCRIPTION
**Pull request by koffes**
Moved the data_fifo library

Signed-off-by: Kristoffer Rist Skøien <kristoffer.skoien@nordicsemi.no>

### CODE REVIEW

File: doc/nrf/libraries/others/data_fifo.rst [hunk 1]
```diff
@@ -0,0 +1,29 @@
+.. _lib_data_fifo:
+
+Data FIFO
+##############
```
Comment (greg-fer) [inline doc/nrf/libraries/others/data_fifo.rst:4]:
```suggestion
Data FIFO
#########
```

File: doc/nrf/libraries/others/data_fifo.rst [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+.. contents::
+   :local:
+   :depth: 2
+
+This library combines the Zephyr memory slab and message queue mechanisms.
+The purpose is to be able to allocate a memory slab, use it, and signal to a receiver when the
+write operation has completed.
+The reader can then read and free the memory slab when done.
+For more information, see `API documentation`_.
```
Comment (greg-fer) [inline doc/nrf/libraries/others/data_fifo.rst:14]:
```suggestion
This library combines the Zephyr memory slab and message queue mechanisms.
The purpose is to be able to allocate a memory slab, use it, and signal to a receiver when the write operation has completed.
The reader can then read and free the memory slab when done.
For more information, see `API documentation`_.
```

File: doc/nrf/libraries/others/data_fifo.rst [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+To enable the library, set the :kconfig:option:`CONFIG_DATA_FIFO` Kconfig option to ``y`` in the project configuration file :file:`prj.conf`.
+
+API documentation
+*****************
+
+| Header file: :file:`include/data_fifo.h`
+| Source file: :file:`lib/tone/data_fifo.c`
+
+.. doxygengroup:: tone_gen
```
Comment (greg-fer) [inline doc/nrf/libraries/others/data_fifo.rst:27]: This needs updating to the new library's doxygen group.

File: include/data_fifo.h [hunk 4]
```diff
@@ -49,7 +49,7 @@ struct data_fifo {
  * @param data_fifo Pointer to the data_fifo structure
```
Comment (greg-fer) [inline include/data_fifo.h:49]: This header is missing the opening definitions for doxygen groups. Please add them using the headers of previously added libraries as reference.

(Note: Closing tags are needed at the end of the file too.)

**Review by greg-fer — CHANGES_REQUESTED** [review]

**Comment by lemrey** [comment]
PR NUMBER 10.000 🥳

File: include/data_fifo.h [hunk 5]
```diff
@@ -74,7 +82,7 @@ int data_fifo_pointer_first_vacant_get(struct data_fifo *data_fifo, void **data,
  * @param size Number of bytes written. Must be equal to or smaller
  *		than the block size max.
  *
- * @retval 0		Block has been sumbitted to the message queue.
+ * @retval 0		Block has been submitted to the message queue.
  * @retval -ENOMEM	size is larger than block size max.
  * @retval -EINVAL	Supplied size is zero
```
Comment (erikrobstad) [inline include/data_fifo.h:87]:
```suggestion
 * @retval -EINVAL	Supplied size is zero.
```

File: tests/lib/data_fifo/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,13 @@
+#
+# Copyright (c) 2018 Nordic Semiconductor ASA
```
Comment (erikrobstad) [inline tests/lib/data_fifo/CMakeLists.txt:2]: 2018 = 2023?

File: lib/data_fifo/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+zephyr_library()
+zephyr_library_sources(
+	data_fifo.c
+)
+
+zephyr_include_directories(.)
```
Comment (erikrobstad) [inline lib/data_fifo/CMakeLists.txt:12]: There are no .h files here?

File: lib/data_fifo/Kconfig [hunk 8]
```diff
@@ -0,0 +1,19 @@
+# Continuous Array creator
```
Comment (erikrobstad) [inline lib/data_fifo/Kconfig:1]: We dont have this headline in e.g. lib/tone

**Review by erikrobstad — CHANGES_REQUESTED** [review]

File: lib/data_fifo/CMakeLists.txt (hunk 7 above)
Comment (koffes) [reply lib/data_fifo/CMakeLists.txt:12]: Correct. NCS includes *.h files from a different location.

File: tests/lib/data_fifo/CMakeLists.txt (hunk 6 above)
Comment (koffes) [reply tests/lib/data_fifo/CMakeLists.txt:2]: The tests originate from 2018.

File: lib/data_fifo/CMakeLists.txt (hunk 7 above)
Comment (erikrobstad) [reply lib/data_fifo/CMakeLists.txt:12]: But then "zephyr_include_directories(.)" is not needed?

File: doc/nrf/libraries/others/data_fifo.rst [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+Configuration
+*************
+
+To enable the library, set the :kconfig:option:`CONFIG_DATA_FIFO` Kconfig option to ``y`` in the project configuration file :file:`prj.conf`.
+
+API documentation
+*****************
+
+| Header file: :file:`include/data_fifo.h`
+| Source file: :file:`lib/data_fifo.c`
```
[the remark this answers is not in the document]
Comment (erikrobstad) [reply doc/nrf/libraries/others/data_fifo.rst:24]: I forgot the folder in my previous suggested change

**Review by greg-fer — APPROVED** [review]
Don't forget the entry in the changelog, as this is a new library.

**Review by greg-fer — COMMENTED** [review]
Actually, let me commit a couple of minor edits.

**Review by erikrobstad — APPROVED** [review]

**Review by Frodevan — CHANGES_REQUESTED** [review]
Don't forget to add triggers for the moved files to .github/test-spec.yml

**Review by Frodevan — APPROVED** [review]

**Pull request merged**